lib/hl7/tokenizer.ex

```diff
@@ -5,63 +5,66 @@ defmodule HL7.Tokenizer do
+    tokenize_terminator(rest, original, skip, len, acc, @repetition, copy_fun)
   end
 
-  defp tokenize(<<"^", rest::binary>>, original, skip, len, acc) do
-    tokenize_terminator(rest, original, skip, len, acc, @component)
+  defp tokenize(<<"^", rest::binary>>, original, skip, len, acc, copy_fun) do
+    tokenize_terminator(rest, original, skip, len, acc, @component, copy_fun)
   end
 
-  defp tokenize(<<"&", rest::binary>>, original, skip, len, acc) do
-    tokenize_terminator(rest, original, skip, len, acc, @sub_component)
+  defp tokenize(<<"&", rest::binary>>, original, skip, len, acc, copy_fun) do
+    tokenize_terminator(rest, original, skip, len, acc, @sub_component, copy_fun)
   end
 
-  defp tokenize(<<"\r", rest::binary>>, original, skip, len, acc) do
-    tokenize_terminator(rest, original, skip, len, acc, @carriage_return)
+  defp tokenize(<<"\r", rest::binary>>, original, skip, len, acc, copy_fun) do
+    tokenize_terminator(rest, original, skip, len, acc, @carriage_return, copy_fun)
   end
 
-  defp tokenize(<<"\n", rest::binary>>, original, skip, len, acc) do
-    tokenize_terminator(rest, original, skip, len, acc, @carriage_return)
+  defp tokenize(<<"\n", rest::binary>>, original, skip, len, acc, copy_fun) do
+    tokenize_terminator(rest, original, skip, len, acc, @carriage_return, copy_fun)
   end
 
-  defp tokenize(<<_char::binary-size(1), rest::binary>>, original, skip, len, acc) do
-    tokenize(rest, original, skip, len + 1, acc)
+  defp tokenize(<<_char::binary-size(1), rest::binary>>, original, skip, len, acc, copy_fun) do
+    tokenize(rest, original, skip, len + 1, acc, copy_fun)
   end
 
-  defp tokenize("", _original, _skip, 0, acc) do
+  defp tokenize("", _original, _skip, 0, acc, _copy_fun) do
     Enum.reverse(acc)
   end
 
-  defp tokenize("", original, skip, len, acc) do
-    string = binary_part(original, skip, len)
+  defp tokenize("", original, skip, len, acc, copy_fun) do
+    string = copy_fun.(binary_part(original, skip, len))
     Enum.reverse([string | acc])
   end
 
-  defp tokenize_terminator(text, original, skip, 0, acc, terminator) do
-    tokenize(text, original, skip + 1, 0, [terminator | acc])
+  defp tokenize_terminator(text, original, skip, 0, acc, terminator, copy_fun) do
+    tokenize(text, original, skip + 1, 0, [terminator | acc], copy_fun)
   end
 
-  defp tokenize_terminator(text, original, skip, len, acc, terminator) do
-    string = binary_part(original, skip, len)
-    tokenize(text, original, skip + len + 1, 0, [terminator, string | acc])
+  defp tokenize_terminator(text, original, skip, len, acc, terminator, copy_fun) do
+    string = copy_fun.(binary_part(original, skip, len))
+    tokenize(text, original, skip + len + 1, 0, [terminator, string | acc], copy_fun)
   end
 
-  @compile {:inline, tokenize: 5, tokenize_terminator: 6}
+  defp get_copy_fun(true), do: &:binary.copy/1
+  defp get_copy_fun(false), do: &(&1)
```

Comment:
Super nitpick - but this could be `&Function.identity/1` - which may be a hair more friendly to read

Reply:
you character count upper you!
ok i yield :)